File: Java/Concurrency/BoundedQueueDemo.java
package Concurrency;

import java.util.concurrent.*;
import java.util.LinkedList;
import java.util.Queue;

public class BoundedQueueDemo {

    static class BoundedQueue<T> {
        private final Queue<T> queue = new LinkedList<>();
        private final int capacity;

        public BoundedQueue(int capacity) {
            this.capacity = capacity;
        }

        public synchronized void put(T item) throws InterruptedException {
            while (queue.size() == capacity) {          // full — wait for a taker
                System.out.println(Thread.currentThread().getName() + " sees FULL (" + queue.size() + "/" + capacity + "), waiting...");
                wait();
            }
            queue.offer(item);
            System.out.println(Thread.currentThread().getName() + " put: " + item + " [size=" + queue.size() + "]");
            notifyAll();                                  // wake possible waiting takers (and putters — see note below)
        }

        public synchronized T take() throws InterruptedException {
            while (queue.isEmpty()) {                     // empty — wait for a putter
                System.out.println(Thread.currentThread().getName() + " sees EMPTY, waiting...");
                wait();
            }
            T item = queue.poll();
            System.out.println(Thread.currentThread().getName() + " took: " + item + " [size=" + queue.size() + "]");
            notifyAll();
            return item;
        }
    }

    public static void main(String[] args) throws InterruptedException {
        BoundedQueue<String> queue = new BoundedQueue<>(3);   // capacity 3, not 1
        ExecutorService pool = Executors.newFixedThreadPool(4);

        Runnable producer = () -> {
            try {
                for (int i = 0; i < 5; i++) {
                    queue.put(Thread.currentThread().getName() + "-item" + i);
                    Thread.sleep(30);
                }
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }
        };

        Runnable consumer = () -> {
            try {
                for (int i = 0; i < 5; i++) {
                    queue.take();
                    Thread.sleep(60);
                }
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }
        };

        pool.submit(producer);
        pool.submit(producer);
        pool.submit(consumer);
        pool.submit(consumer);

        pool.shutdown();
        pool.awaitTermination(5, TimeUnit.SECONDS);
    }
}

File: Golang/Concurrency/BoundedQueueDemo.go
package main

import (
	"fmt"
	"sync"
	"time"
)

type BoundedQueue struct {
	mu       sync.Mutex
	cond     *sync.Cond
	items    []string
	capacity int
}

func NewBoundedQueue(capacity int) *BoundedQueue {
	q := &BoundedQueue{capacity: capacity}
	q.cond = sync.NewCond(&q.mu)
	return q
}

func (q *BoundedQueue) Put(name, item string) {
	q.mu.Lock()
	defer q.mu.Unlock()
	for len(q.items) == q.capacity { // full — wait for a taker
		fmt.Printf("%s sees FULL (%d/%d), waiting...\n", name, len(q.items), q.capacity)
		q.cond.Wait()
	}
	q.items = append(q.items, item)
	fmt.Printf("%s put: %s [size=%d]\n", name, item, len(q.items))
	q.cond.Broadcast() // wake possible waiting takers (and putters)
}

func (q *BoundedQueue) Take(name string) string {
	q.mu.Lock()
	defer q.mu.Unlock()
	for len(q.items) == 0 { // empty — wait for a putter
		fmt.Printf("%s sees EMPTY, waiting...\n", name)
		q.cond.Wait()
	}
	item := q.items[0]
	q.items = q.items[1:]
	fmt.Printf("%s took: %s [size=%d]\n", name, item, len(q.items))
	q.cond.Broadcast()
	return item
}

func producer(wg *sync.WaitGroup, queue *BoundedQueue, name string) {
	defer wg.Done()
	for i := 0; i < 5; i++ {
		queue.Put(name, fmt.Sprintf("%s-item%d", name, i))
		time.Sleep(30 * time.Millisecond)
	}
}

func consumer(wg *sync.WaitGroup, queue *BoundedQueue, name string) {
	defer wg.Done()
	for i := 0; i < 5; i++ {
		queue.Take(name)
		time.Sleep(60 * time.Millisecond)
	}
}

func main() {
	queue := NewBoundedQueue(3) // capacity 3, not 1

	var wg sync.WaitGroup
	wg.Add(4)
	go producer(&wg, queue, "producer-1")
	go producer(&wg, queue, "producer-2")
	go consumer(&wg, queue, "consumer-1")
	go consumer(&wg, queue, "consumer-2")
	wg.Wait()
}
